src/relay/pass/type_solver.cc

```diff
@@ -289,30 +289,44 @@ class TypeSolver::Unifier : public TypeFunctor<Type(const Type&, const Type&)> {
     const auto* ftn = tn.as<FuncTypeNode>();
     if (!ftn
         || op->arg_types.size() != ftn->arg_types.size()
-        || op->type_params.size() != ftn->type_params.size()
         || op->type_constraints.size() != ftn->type_constraints.size()) {
       return Type(nullptr);
     }
 
+    // without loss of generality, suppose op->type_params.size() >= ftn->type_params.size().
+    if (op->type_params.size() < ftn->type_params.size()) {
+      return VisitType_(ftn, GetRef<FuncType>(op));
+    }
+
     // remap type vars so they match
     Map<TypeVar, Type> subst_map;
-    for (size_t i = 0; i < op->type_params.size(); i++) {
-      subst_map.Set(ftn->type_params[i], op->type_params[i]);
+    tvm::Array<TypeVar> ft_type_params;
+    for (size_t i = 0; i < ftn->type_params.size(); ++i) {
+      subst_map.Set(op->type_params[i], ftn->type_params[i]);
+      ft_type_params.push_back(op->type_params[i]);
+    }
+
+    for (size_t i = ftn->type_params.size(); i < op->type_params.size(); ++i) {
+      subst_map.Set(op->type_params[i], IncompleteTypeNode::make(kType));
     }
 
-    auto ft1 = GetRef<FuncType>(op);
-    auto ft2 = Downcast<FuncType>(Bind(GetRef<FuncType>(ftn), subst_map));
+    FuncType ft = FuncTypeNode::make(op->arg_types,
+                                     op->ret_type,
+                                     ft_type_params,
+                                     op->type_constraints);
+    auto ft1 = Downcast<FuncType>(Bind(ft, subst_map));
+    auto ft2 = GetRef<FuncType>(ftn);
 
     Type ret_type = Unify(ft1->ret_type, ft2->ret_type);
 
     std::vector<Type> arg_types;
-    for (size_t i = 0; i < ft1->arg_types.size(); i++) {
+    for (size_t i = 0; i < ft2->arg_types.size(); ++i) {
```

Comment:
what's the relation between `ft1->arg_types.size()` and `ft2->arg_types.size()`

Reply:
I imagine it has to do with the assumption "op->type_params.size() >= ftn->type_params.size()" (so `ft2->arg_types.size()` would be less than or equal to `ft1->arg_types.size()`). Is that right @MarisaKirisame ? There should probably be an additional comment.